tempesta_fw/http.c

```diff
@@ -1346,15 +1417,10 @@ tfw_http_conn_msg_alloc(TfwConn *conn)
 	if (TFW_CONN_TYPE(conn) & Conn_Clnt) {
 		TFW_INC_STAT_BH(clnt.rx_messages);
 	} else {
-		TfwHttpReq *req;
-		TfwSrvConn *srv_conn = (TfwSrvConn *)conn;
-
-		spin_lock(&srv_conn->fwd_qlock);
-		req = list_first_entry_or_null(&srv_conn->fwd_queue,
-					       TfwHttpReq, fwd_list);
-		spin_unlock(&srv_conn->fwd_qlock);
-		if (req && (req->method == TFW_HTTP_METH_HEAD))
-			hm->flags |= TFW_HTTP_VOID_BODY;
+		if (unlikely(tfw_http_resp_bind_req(hm))) {
```

Comment:
While this comment does apply to this specific line of code, this is more of a general comment.

My concern is that I don't completely understand the implication of a possible break in communication between Tempesta and a back end server. The code works the way that it *removes* the paired request from the forwarding queue as soon as the response starts coming in. If the back end is unable to transmit a full HTTP response due to some reason, wouldn't it attempt to re-send the response again later? It the paired request is already removed from the forwarding queue, that will break the pairing of requests and responses.

I would say that early binding should not remove the corresponding request from the forwarding queue just yet. It's simply a linkage. When the response is parsed, only then the request should be removed.

What do you think?

Reply:
Yes, I've missed that scenario. 

> I would say that early binding should not remove the corresponding request from the forwarding queue just yet. It's simply a linkage. When the response is parsed, only then the request should be removed.

I'm agree here. I'll fix that.